### PR TITLE
feat(ecmascript): Array iterator objects

### PR DESCRIPTION
**Pull request by andreubotella**

### CODE REVIEW

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_iterator_objects/array_iterator_prototype.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            }
+            // 3. If kind is value, then
+            ArrayIteratorKind::Value => {
+                // 1. Let elementKey be ! ToString(indexNumber).
+                // 2. Let elementValue be ? Get(array, elementKey).
+                // a. Let result be elementValue.
+                match array {
+                    Object::Array(array) if agent[array].object_index.is_none() => {
+                        let idx = usize::try_from(index.into_i64()).unwrap();
+                        array.as_slice(agent)[idx].unwrap_or(Value::Undefined)
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_iterator_objects/array_iterator_prototype.rs:97]: **issue:** If the slot contains None or is out of bounds, then we need to go into the prototype chain or call possibly call a getter.

Checking the slot being Some can also be done even if we have a backing object.

The internal_get method of Array could check that for indexed properties directly actually.

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs [hunk 2]
```diff
@@ -573,8 +575,18 @@ impl ArrayPrototype {
         Ok(o.into_value())
     }
 
-    fn entries(_agent: &mut Agent, _this_value: Value, _: ArgumentsList) -> JsResult<Value> {
-        todo!()
+    fn entries(agent: &mut Agent, this_value: Value, _: ArgumentsList) -> JsResult<Value> {
+        // 1. Let O be ? ToObject(this value).
+        let Value::Array(o) = this_value else {
+            return Err(
+                agent.throw_exception_with_static_message(ExceptionType::TypeError, "blargh")
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs:582]: **praise:** blargh indeed

File: nova_vm/src/heap.rs [hunk 3]
```diff
@@ -153,6 +155,7 @@ impl Heap {
         let mut heap = Heap {
             array_buffers: Vec::with_capacity(1024),
             arrays: Vec::with_capacity(1024),
+            array_iterators: Vec::with_capacity(1024),
```
Comment (aapoalas) [inline nova_vm/src/heap.rs:158]: **nitpick:** Probably unnecessarily large default size.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Wrong logic for array indexed property access. Otherwise looks really good, thank you! <3

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs (hunk 2 above)
Comment (andreubotella) [reply nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_prototype.rs]: I also didn't realize that I changed the implementation of `next` to allow non-array this values but didn't change the array prototype methods to allow arbitrary objects.

File: nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_iterator_objects/array_iterator.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    /// This is only for Value discriminant creation.
+    pub(crate) const fn _def() -> Self {
+        Self(ArrayIteratorIndex::from_u32_index(0))
+    }
+
+    pub(crate) fn get_index(self) -> usize {
+        self.0.into_index()
+    }
+
+    pub(crate) fn from_array(agent: &mut Agent, array: Object, kind: ArrayIteratorKind) -> Self {
```
Comment (aapoalas) [inline nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_iterator_objects/array_iterator.rs:35]: **nitpick:** Should probably be "from_object".

**Review by aapoalas — APPROVED** [review]

**Pull request merged**